src/core/graph.js

```diff
@@ -889,6 +898,55 @@ export class Graph {
     return result;
   }
 
+  /**
+   * Create a new graph, in which some nodes have been contracted together.
+   *
+   * contract takes a list of Contractions, each of which specifies a
+   * replacement node, and a list of old node addresses to map onto the new
+   * node. A new graph will be returned where the new node is added, none of
+   * the old nodes are present, and every edge incident to one of the old nodes
+   * has been re-written so that it is incident to the new node instead.
+   *
+   * If the same node addresses is "old" for several contractions, all incident
+   * edges will be re-written to connect to whichever contraction came last.
+
+   * The original Graph is not mutated.
+   *
+   * contract runs in O(n+e+k), where `n` is the number of nodes, `e` is the
+   * number of edges, and `k` is the number of contractions. If needed, we can
+   * improve the peformance by mutating the original graph instead of creating
+   * a new one.
+   */
+  contract(contractions: $ReadOnlyArray<GraphContraction>): Graph {
```

Comment:
This doesn’t handle chained contractions, right? It looks like both

```js
graph.contract([
  {old: [a.address], new: b},
  {old: [b.address], new: c},
]);
```

and

```js
graph.contract([
  {old: [b.address], new: c},
  {old: [a.address], new: b},
]);
```

will turn `a` into two nodes `b` and `c`, with incident edges remapped
to `b`, and `c` an isolated vertex?

The semantics that you want, I think, are that you imagine applying the
contractions to the graph one at a time, but when you remap nodes and
edges in the graph you also remap later entries in the contractions
list. This conservatively extends the existing “take-last” semantics to
also handle chains.

Reply:
I agree that the semantics you describe make sense. I spent at least 10 minutes thinking about how to get these semantics without either significantly complicating the implementation (keeping a backward map pointing every replacement to its origins), or making the runtime much worse `O(n+e+k) -> O(nk + ek)`, and didn't find any good solutions. Since we don't actually have any need for chained contractions, I don't think it's worth the effort to implement and test the more complicated behavior. Therefore, I instead implemented logic to detect any chained contractions and throw an error.

I left a (commented) block of code with a test for the chained contractions, so if we decide we need it later, we'll have a bit of a head start, test-driven-development style. If you see an elegant way to implement contraction chaining, feel free to do so in a followon PR. (I considered skipping the test instead of commenting it, which would decrease the probability of code rot, but I feel like there's some benefit to having 0 skipped tests--it means that if we start seeing test skips in jest, it indicates some new kind of technical debt that we're accumulating.)

Reply:
The approach that comes immediately to mind is to use a union-find
structure to accumulate the connected components, plus an extra forward
pass to resolve roots to their final node names. This should be nearly
linear time (up to α(·)), but I’ll grant that it’s not straightforward
given that there’s no union-find structure in the “standard library”.

Detecting and failing sounds good to me.